### ISSUE TITLE
Core/SAI: (feature request) `SMART_ACTION` types for sending worldstates to the client

### ISSUE THREAD

**Issue by ihm-tswow** [TrinityCore/TrinityCore#26907]
### Description

_tldr: we're adding two new `SMART_ACTION` types for custom content to tswow, forked off TrinityCore. If you want to I can PR the same actions here and we can agree on using the same action ID to futureproof it, since at least AzerothCore will also add it. If you don't want that kind of stuff in your own sai, you can safely close this issue because then we'd never clash in the first place._

**The problem**
Currently, there is no good way to send worldstates for use in gossip texts without core scripting. This is a bit unfortunate for things like war effort-style custom game events, where you'd probably want to update that along with sending the gossip menu: 
![image](https://user-images.githubusercontent.com/76849026/132997183-623e11f3-daf2-4767-8436-3f43a71c11b6.png)

The solution we've made are two new `SMART_ACTION` types: 

### `SMART_ACTION_SEND_WORLDSTATE` 
Calls `Player::SendUpdateWorldState` either with a value stored in the server (`worldstates` table in tc) or a hardcoded value. Can be used for sending arbitrary server worldstates to the client when opening a gossip menu.

- param1 decides the worldstate id to send
- param2=0 if it's read from the worldserver table, or =1 if it's an immediate value. Can be extended with whatever.
- param3 is the immediate value to send if param2==0, otherwise ignored. 

### `SMART_ACTION_SEND_EVENT_STATE`
Calls `GameEventMgr::SendWorldStateUpdate`, which sends all the condition worldstates to a target player. Can be used for war effort-style gossips as shown above

- param1 = the `game_event` to send worldstates for. 

We're going to be adding this to tswow anyways so thought we'd just check if you also want the patch, would just be annoying if we both ended up implementing it in the end with different ids. Just let me know if you want it and I can make a pr.

### Expected behaviour

Being able to send the `max_world_state_field` and `done_world_state_field` columns in `game_event_condition` without scripting the core, since that's all they're used for.

### Steps to reproduce the problem

1. Try to make custom war effort
2. No SAI action to send worldstates
3. Cry because you don't want to corescript

### Branch

3.3.5

### TC rev. hash/commit

a18b19aeceea5cc1c3c6ce30b4790cc81bf10bcb (yes, I checked it wasn't added since our latest merge :P)

### Operating system

Windows 10 x64, Debian 10 x64, Ubuntu 20.04.3 x64

### Custom changes

So many it'd make you cry. That's why I ask if you're interested before making a pr to the clean thing.

**Comment by Faq** [comment]
As here mentioned custom event, then this more likely for https://github.com/TrinityCore/TrinityCoreCustomChanges

**Comment by Shauren** [comment]
Both of these actions are based on current state of worldstate support in TC however there is a future change which should make both of them unneccessary (currently in cata fork - https://github.com/The-Cataclysm-Preservation-Project/TrinityCore/commit/5fd77a3b284fc7ab9abf2ea201b40016e197fba5)

**Comment by ihm-tswow** [comment]
> Both of these actions are based on current state of worldstate support in TC however there is a future change which should make both of them unneccessary (currently in cata fork - [The-Cataclysm-Preservation-Project@5fd77a3](https://github.com/The-Cataclysm-Preservation-Project/TrinityCore/commit/5fd77a3b284fc7ab9abf2ea201b40016e197fba5))

This is a global broadcast, so it _can_ still be combined with actions in case you don't want to globally broadcast every worldstate change stored in that table (that can become a lot of packets everytime a player logs in and needs to synchronize / if a worldstate changes a lot). Worldstates are often just displayed in a single area or used with things like gossips, so most often it's enough to send them when the client actually needs them.

A minor additional point is that `game_event` worldstates aren't actually currently stored as real worldstates since they have their own table (i guess to store them closer to the actual event in memory?), but they can of course be broadcast / synchronized the same way.

For us, it doesn't matter too much which strategy you decide on, and we can easily work around it if you decide to start automatically broadcasting them. Would just be good to use the same ids _if_ we use the same /similar sai actions 😄

**Comment by Ovahlord** [comment]
Realm wide world states are suposed to be sent to every player, as well as broadcasting map wide world states to all players on the map. All that is based on sniff research, so sending it to just one player for a quest is straightup a hack.

**Comment by ihm-tswow** [comment]
> Realm wide world states are suposed to be sent to every player, as well as broadcasting map wide world states to all players on the map. All that is based on sniff research, so sending it to just one player for a quest is straightup a hack.

Thanks for the clarification! I'm not active with emulation anymore so I don't follow the research (and never worked with worldstates when I was). For tswow, we will stick to these actions since we only use it for custom and in that area allows us some more flexibility, but it's good to know there is no reason to add conflicting actions to the emulator itself.
